src/strategies/types.ts

```diff
@@ -5,61 +5,59 @@ import * as AzureADStrategy from 'passport-azure-ad';
 import {IAuthClient, IAuthUser} from '../types';
 
 export namespace VerifyFunction {
-  export interface OauthClientPasswordFn {
-    (
-      clientId: string,
-      clientSecret: string,
-      req?: Request,
-    ): Promise<IAuthClient | null>;
+  export interface OauthClientPasswordFn<T = IAuthClient> {
+    (clientId: string, clientSecret: string, req?: Request): Promise<T | null>;
   }
 
-  export interface LocalPasswordFn {
-    (
-      username: string,
-      password: string,
-      req?: Request,
-    ): Promise<IAuthUser | null>;
+  export interface LocalPasswordFn<T = IAuthUser> {
+    (username: string, password: string, req?: Request): Promise<T | null>;
   }
 
-  export interface BearerFn {
-    (token: string, req?: Request): Promise<IAuthUser | null>;
+  export interface BearerFn<T = IAuthUser> {
+    (token: string, req?: Request): Promise<T | null>;
   }
 
-  export interface ResourceOwnerPasswordFn {
+  export interface ResourceOwnerPasswordFn<T = IAuthClient, S = IAuthUser> {
     (
       clientId: string,
       clientSecret: string,
       username: string,
       password: string,
       req?: Request,
-    ): Promise<{client: IAuthClient; user: IAuthUser} | null>;
+    ): Promise<{client: T; user: S} | null>;
   }
 
-  export interface GoogleAuthFn {
+  export interface GoogleAuthFn<T = IAuthUser> {
     (
       accessToken: string,
       refreshToken: string,
       profile: GoogleStrategy.Profile,
       cb: GoogleStrategy.VerifyCallback,
       req?: Request,
-    ): Promise<IAuthUser | null>;
+    ): Promise<T | null>;
   }
 
-  export interface AzureADAuthFn {
+  export interface AzureADAuthFn<T = IAuthUser> {
```

Comment:
Can this extend interface `GenericAuthFn` ?
https://www.logicbig.com/tutorials/misc/typescript/interface-extending-interfaces.html

Reply:
Done, extended GenericAuthFn to ensure Promise<T> return type